File: typescript/test/normalizer.test.ts
import { describe, expect, it } from "vitest";
import { normalizeRecord } from "../src/normalizer.js";
import type { FieldDefinition, FormDefinition } from "../src/definition.js";

const field = (name: string, normalize: string[]): FieldDefinition => ({
  field: name,
  label: name,
  type: "text",
  required: false,
  readOnly: false,
  validators: [],
  options: [],
  normalize,
  config: {},
});

describe("normalizeRecord", () => {
  const form: FormDefinition = {
    sections: [{ columns: 1, fields: [field("code", ["toHankaku", "trim"]), field("name", [])] }],
  };

  it("applies converter chains per field, leaves others untouched", () => {
    const out = normalizeRecord(form, { code: "　ＡＢ１２　", name: "　x　" });
    expect(out.code).toBe("AB12");
    expect(out.name).toBe("　x　");
  });
});

File: flutter/packages/hatake_core/test/normalizer_test.dart
import 'package:hatake_core/hatake_core.dart';
import 'package:test/test.dart';

void main() {
  test('FormNormalizer applies converter chains per field', () {
    const form = FormDefinition(sections: [
      SectionDefinition(fields: [
        FieldDefinition(
          field: 'code',
          label: 'コード',
          normalize: ['toHankaku', 'trim'],
        ),
        FieldDefinition(field: 'name', label: '名前'), // no normalize
      ])
    ]);

    final out = FormNormalizer().normalize(form, {
      'code': '　ＡＢ１２　',
      'name': '　x　',
    });

    expect(out['code'], 'AB12');
    expect(out['name'], '　x　'); // untouched (no normalize declared)
  });

  test('leaves absent fields alone', () {
    const form = FormDefinition(sections: [
      SectionDefinition(fields: [
        FieldDefinition(field: 'code', label: 'コード', normalize: ['toHankaku']),
      ])
    ]);
    final out = FormNormalizer().normalize(form, {'other': 1});
    expect(out.containsKey('code'), isFalse);
    expect(out['other'], 1);
  });
}
